File: server_play_akka_scala/app/domain/logic/LunchResolverKrauthof.scala
package domain.logic

import java.io.FileNotFoundException
import java.net.URL
import java.time.format.DateTimeFormatter
import java.time.{DayOfWeek, LocalDate}

import domain.models.{LunchOffer, LunchProvider}
import domain.util.{PDFTextGroupStripper, TextLine}
import domain.util.Html
import org.apache.pdfbox.pdmodel.PDDocument
import org.joda.money.{CurrencyUnit, Money}
import util.PlayLogging

import scala.concurrent.ExecutionContext.Implicits.global
import scala.concurrent.Future
import scala.util.matching.Regex

class LunchResolverKrauthof(dateValidator: DateValidator) extends LunchResolver with PlayLogging {

  sealed abstract class PdfSection(val sectionStartPattern: String, val order: Int)

  object PdfSection {
    case object HEADER extends PdfSection("<<Header>>", 0)
    case object MONTAG extends PdfSection("Montag", 0)
    case object DIENSTAG extends PdfSection("Dienstag", 1)
    case object MITTWOCH extends PdfSection("Mittwoch", 2)
    case object DONNERSTAG extends PdfSection("Donnerstag", 3)
    case object FREITAG extends PdfSection("Freitag", 4)
    case object FOOTER extends PdfSection("<<Footer>>", 0)

    val weekdaysValues = List[PdfSection](MONTAG, DIENSTAG, MITTWOCH, DONNERSTAG, FREITAG)
    // TODO: improve with macro, see https://github.com/d6y/enumeration-examples & http://underscore.io/blog/posts/2014/09/03/enumerations.html
    val values = weekdaysValues // ++ List(SALAT_DER_WOCHE, FOOTER)
  }

  case class OfferRow(name: String, priceOpt: Option[Money]) {
    def merge(otherRow: OfferRow): OfferRow = {
      val newName = Seq(name, otherRow.name).filter(!_.isEmpty).mkString(" ")
      OfferRow(newName, priceOpt.orElse(otherRow.priceOpt))
    }

    def isValid = !name.isEmpty && priceOpt.isDefined
  }

  implicit class RegexContext(sc: StringContext) {
    def r = new Regex(sc.parts.mkString, sc.parts.tail.map(_ => "x"): _*)
  }

  override def resolve: Future[Seq[LunchOffer]] =
    Future { resolvePdfLinks(new URL("https://www.daskrauthof.de/karte")) }
      .flatMap(pdfPaths => resolveFromPdfs(pdfPaths))

  private[logic] def resolvePdfLinks(htmlUrl: URL): Seq[String] = {
    val siteAsXml = Html.load(htmlUrl)

    val links = (siteAsXml \\ "a").map(_ \@ "href")
    links.filter(_ matches """.*/KRAUTHOF.*.pdf""").toList
  }

  private def resolveFromPdfs(pdfPaths: Seq[String]): Future[Seq[LunchOffer]] = {
    val listOfFutures = pdfPaths.map(pdfPath =>
      Future { resolveFromPdf(new URL(pdfPath)) })
    Future.sequence(listOfFutures).map(listOfLists => listOfLists.flatten)
  }

  private[logic] def resolveFromPdf(pdfUrl: URL): Seq[LunchOffer] = {
    val pdfContent = extractPdfContent(pdfUrl)

    parseMondayFromContent(pdfContent)
      .filter(dateValidator.isValid)
      .map(resolveFromPdfContent(pdfContent, _))
      .getOrElse(Nil)
  }

  private[logic] def resolveFromPdfContent(pdfContent: Seq[String], monday: LocalDate): Seq[LunchOffer] = {
    var rows = Seq[OfferRow]()
    var currentRow: Option[OfferRow] = None

    def finishOffer() = {
      currentRow.foreach(row => if (row.isValid) rows :+= row)
      currentRow = None
    }

    for (line <- pdfContent) {
      line.trim match {
        case r"""(.+)$text(\d{1,}[.,]\d{2})$priceString *€\*""" =>
          finishOffer()
          currentRow = Some(OfferRow(parseName(text), parsePrice(priceString)))
        case "" | r""".*inklusive Tagesgetränk.*""" =>
          finishOffer()
        case text =>
          val thisRow = OfferRow(parseName(text), None)
          currentRow = currentRow match {
            case Some(offer) => Some(offer.merge(thisRow))
            case None => Some(thisRow)
          }
      }
    }

    PdfSection.weekdaysValues.flatMap(
      weekday => rows.map(
        row => LunchOffer(0, row.name, monday.plusDays(weekday.order), row.priceOpt.get, LunchProvider.DAS_KRAUTHOF.id)))
  }

  private def extractPdfContent(pdfUrl: URL): Seq[String] = {
    var optPdfDoc: Option[PDDocument] = None
    var pdfContent = Seq[TextLine]()

    try {
      optPdfDoc = Some(PDDocument.load(pdfUrl))
      for (pdfDoc <- optPdfDoc) {
        val stripper = new PDFTextGroupStripper
        pdfContent = stripper.getTextLines(pdfDoc)
      }
    } catch {
      case fnf: FileNotFoundException => logger.error(s"file $pdfUrl not found")
      case t: Throwable => logger.error(s"Fehler beim Einlesen von $pdfUrl", t)
    } finally {
      optPdfDoc.foreach(_.close())
    }
    pdfContent.map(_.toString)
  }

  private[logic] def parseMondayFromContent(lines: Seq[String]): Option[LocalDate] = {
    // alle Datumse aus PDF ermitteln
    val days = lines.flatMap(line => parseDay(line.toString))
    val mondays = days.map(_.`with`(DayOfWeek.MONDAY))
    // den Montag der am häufigsten verwendeten Woche zurückgeben
    mondays match {
      case Nil => None
      case _ => Option(mondays.groupBy(identity).maxBy(_._2.size)._1)
    }
  }

  private def parseDay(dayString: String): Option[LocalDate] = dayString match {
    case r""".*(\d{2}.\d{2}.\d{4})$dayString.*""" => parseLocalDate(dayString, "dd.MM.yyyy")
    case r""".*(\d{2}.\d{2}.\d{2})$dayString.*""" => parseLocalDate(dayString, "dd.MM.yy")
    case r""".*(\d{2}.\d{2})$dayString.*""" =>
      val yearToday = LocalDate.now.getYear
      val year = if (LocalDate.now.getMonthValue == 12 && dayString.endsWith("01")) yearToday + 1 else yearToday
      parseLocalDate(dayString + "." + year.toString, "dd.MM.yyyy")
    case _ => None
  }

  private def parseLocalDate(dateString: String, dateFormat: String): Option[LocalDate] =
    try {
      Some(LocalDate.from(DateTimeFormatter.ofPattern(dateFormat).parse(dateString)))
    } catch {
      case exc: Throwable => None
    }

  private def parseName(text: String): String = text.trim
    .replaceAll("  ", " ")
    .replaceAll("–", "-")
    .replaceAll(""" *[\│\|] *""", ", ")
    .replaceAll(" I ", ", ")
    .replaceAll(" *,", ",")
    .trim

  /**
   * Erzeugt ein Money-Objekt (in EURO) aus dem Format "*0,00*"
   *
   * @param priceString String im Format "*0,00*"
   * @return
   */
  private def parsePrice(priceString: String): Option[Money] = priceString match {
    case r""".*(\d{1,})$major[.,](\d{2})$minor.*""" => Some(Money.ofMinor(CurrencyUnit.EUR, major.toInt * 100 + minor.toInt))
    case _ => None
  }

}

File: src/main/kotlin/lunchbox/domain/logic/LunchResolverKrauthof.kt
package lunchbox.domain.logic

import lunchbox.domain.models.LunchOffer
import lunchbox.domain.models.LunchProvider
import lunchbox.domain.models.LunchProvider.DAS_KRAUTHOF
import lunchbox.util.html.Html
import lunchbox.util.pdf.PdfTextGroupStripper
import lunchbox.util.pdf.TextLine
import org.apache.pdfbox.pdmodel.PDDocument
import org.joda.money.CurrencyUnit
import org.joda.money.Money
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Component
import java.io.FileNotFoundException
import java.net.URL
import java.time.DayOfWeek
import java.time.LocalDate
import java.time.format.DateTimeFormatter

@Component
class LunchResolverKrauthof(
  val dateValidator: DateValidator
) : LunchResolver {

  private val logger = LoggerFactory.getLogger(javaClass)

  override val provider: LunchProvider = DAS_KRAUTHOF

  override fun resolve(): List<LunchOffer> {
    val pdfLinks = resolvePdfLinks(URL("https://www.daskrauthof.de/karte"))
    return resolveFromPdfs(pdfLinks)
  }

  fun resolvePdfLinks(htmlUrl: URL): List<String> {
    val site = Html.load(htmlUrl)

    val links = site.select("a").map { it.attr("href") }
    return links.filter { it.matches(Regex(""".*/KRAUTHOF.*.pdf""")) }
  }

  fun resolveFromPdfs(pdfPaths: List<String>): List<LunchOffer> =
    // TODO: parallelize
    pdfPaths.map { resolveFromPdf(URL(it)) }.flatten()

  fun resolveFromPdf(pdfUrl: URL): List<LunchOffer> {
    val pdfContent = extractPdfContent(pdfUrl)

    val monday = parseMondayFromContent(pdfContent)
    if (monday == null || !dateValidator.isValid(monday))
      return emptyList()
    return resolveFromPdfContent(pdfContent, monday)
  }

  private fun resolveFromPdfContent(pdfContent: List<String>, monday: LocalDate): List<LunchOffer> {
    val rows = mutableListOf<OfferRow>()
    var currentRow: OfferRow? = null

    fun finishOffer() {
      currentRow?.let { if (it.isValid()) rows.add(it) }
      currentRow = null
    }

    for (line in pdfContent.map { it.trim() }) {
      val matchOffer = Regex("""(.+)(\d+[.,]\d{2}) *€\*""").find(line)
      if (matchOffer != null) {
        val (text, priceString) = matchOffer.destructured
        finishOffer()
        currentRow = OfferRow(parseName(text), parsePrice(priceString))
        continue
      }

      if (line.isEmpty()) {
        finishOffer()
        continue
      }

      val matchLastLine = Regex(""".*inklusive Tagesgetränk.*""").find(line)
      if (matchLastLine != null) {
        finishOffer()
        continue
      }

      val thisRow = OfferRow(parseName(line), null)
      currentRow = currentRow?.merge(thisRow) ?: thisRow
    }

    val mondayOffers = rows.map { row ->
      LunchOffer(0, row.name, monday, row.price!!, provider.id)
    }

    // an allen Wochentagen gibt es das selbe Angebot
    return LongRange(0, 4).flatMap { weekdayLong ->
      mondayOffers.map { it.copy(day = monday.plusDays(weekdayLong)) }
    }
  }

  private fun extractPdfContent(pdfUrl: URL): List<String> {
    var pdfDoc: PDDocument? = null
    var pdfContent = emptyList<TextLine>()

    try {
      pdfDoc = PDDocument.load(pdfUrl)
      if (pdfDoc != null) {
        val stripper = PdfTextGroupStripper()
        pdfContent = stripper.getTextLines(pdfDoc)
      }
    } catch (fnf: FileNotFoundException) {
      logger.error("file $pdfUrl not found")
    } catch (t: Throwable) {
      logger.error("Fehler beim Einlesen von $pdfUrl", t)
    } finally {
      pdfDoc?.close()
    }
    return pdfContent.map { it.toString() }
  }

  private fun parseMondayFromContent(lines: List<String>): LocalDate? {
    // alle Datumse aus PDF ermitteln
    val days = lines.mapNotNull { parseDay(it) }
    val mondays = days.map { it.with(DayOfWeek.MONDAY) }

    // den Montag der am häufigsten verwendeten Woche zurückgeben
    return mondays
      .groupBy { it }
      .maxBy { it.value.size }
      ?.key
  }

  private fun parseDay(dayString: String): LocalDate? {
    Regex(""".*(\d{2}.\d{2}.\d{4}).*""").find(dayString)?.let {
      val (dateString) = it.destructured
      return parseLocalDate(dateString, "dd.MM.yyyy")
    }

    Regex(""".*(\d{2}.\d{2}.\d{2}).*""").find(dayString)?.let {
      val (dateString) = it.destructured
      return parseLocalDate(dateString, "dd.MM.yy")
    }

    Regex(""".*(\d{2}.\d{2}).*""").find(dayString)?.let {
      val (dateString) = it.destructured
      val yearToday = LocalDate.now().year
      val year =
        if (LocalDate.now().monthValue == 12 && dateString.endsWith("01"))
          yearToday + 1
        else
          yearToday
      return parseLocalDate("$dateString.$year", "dd.MM.yyyy")
    }

    return null
  }

  private fun parseLocalDate(dateString: String, dateFormat: String): LocalDate? =
    try {
      LocalDate.from(DateTimeFormatter.ofPattern(dateFormat).parse(dateString))
    } catch (exc: Throwable) {
      null
    }

  private fun parseName(text: String): String =
    text.trim()
      .replace("  ", " ")
      .replace("–", "-")
      .replace(Regex(""" *[│|] *"""), ", ")
      .replace(" I ", ", ")
      .replace(" *,", ",")
      .trim()

  /**
   * Erzeugt ein Money-Objekt (in EURO) aus dem Format "*0,00*"
   *
   * @param priceString String im Format "*0,00*"
   * @return
   */
  private fun parsePrice(priceString: String): Money? {
    val regex = Regex(""".*(\d+)[.,](\d{2}).*""")
    val matchResult = regex.find(priceString)
      ?: return null
    val (major, minor) = matchResult.destructured
    return Money.ofMinor(CurrencyUnit.EUR, major.toLong() * 100 + minor.toLong())
  }

  data class OfferRow(
    val name: String,
    val price: Money?
  ) {
    fun merge(otherRow: OfferRow): OfferRow {
      val newName = listOf(name, otherRow.name).filter { it.isNotEmpty() }.joinToString(" ")
      return OfferRow(newName, price ?: otherRow.price)
    }

    fun isValid() = name.isNotEmpty() && price != null
  }
}
